### DISCUSSION TITLE
put tryCatchLog in the knitr evaluate.inline hook

### DISCUSSION THREAD

**Discussion by th-tsai** [rbind/yihui#1478] [Q&A]
I am recently working on a task to catch any message (TRACE, DEBUG, INFO, WARN, ERROR, or FATAL) in the rmarkdown while rendering it.

My first idea was to put render inside tryCatchLog like `tryCatchLog(render("my.rmd"))`. However, this method cannot get any message within the rmd.

Thus, I came up with the second idea. I try to insert the tryCatchLog in the evaluate.inline knitr::hook setting. I expect every line of code will be evaluated under tryCatchLog.

```{r}
knitr::knit_hooks$set( evaluate.inline = 
    function (code, envir = knit_global()) {
        v = withVisible(tryCatchLog(eval(parse_only(code), envir = envir)))  # add tryCatchLog here 
        if (v$visible) 
            knit_print(v$value, inline = TRUE, options = opts_chunk$get())
    }
)
```
Nevertheless, this doesn't work alongside what I expect neither. It seems like the file is knit as usual (no tryCatchLog in the evaluate.inline).

Is that I misunderstand the mechanism of evaluate.inline?

I appreciate any comment or explanation!

**Comment by yihui** [comment]
The `evaluate.inline` hook only works for inline R expressions (`` `r ` ``). I think you need to insert your function in both `evaluate.inline` and `evaluate`:

- https://github.com/yihui/knitr/blob/039b6cfa45dea4ece31f2bcb65ed3b34b1330daa/R/hooks.R#L13-L17
- https://github.com/yihui/knitr/blob/039b6cfa45dea4ece31f2bcb65ed3b34b1330daa/R/block.R#L250

> My first idea was to put render inside tryCatchLog like `tryCatchLog(render("my.rmd"))`. However, this method cannot get any message within the rmd.

To make that work, you must use the chunk options `error = FALSE`, `message = NA`, and `warning = NA`. If you can't set them inside the Rmd document, you may set

```r
options(knitr.chunk.error = FALSE, knitr.chunk.message = NA, knitr.chunk.warning = NA)
```

before calling `rmarkdown::render()`.
